storage/src/main/java/org/apache/kafka/server/log/internals/TxnMetadata.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.server.log.internals;
+
+import java.util.Objects;
+import java.util.OptionalLong;
+
+public final class TxnMetadata {
+    public final long producerId;
+    public final LogOffsetMetadata firstOffset;
+    public OptionalLong lastOffset;
+
+    public TxnMetadata(long producerId,
+                       LogOffsetMetadata firstOffset,
+                       OptionalLong lastOffset) {
+        this.producerId = producerId;
+        this.firstOffset = firstOffset;
+        this.lastOffset = lastOffset;
+    }
+    public TxnMetadata(long producerId, long firstOffset) {
+        this(producerId, new LogOffsetMetadata(firstOffset));
+    }
+
+    public TxnMetadata(long producerId, LogOffsetMetadata firstOffset) {
+        this(producerId, firstOffset, OptionalLong.empty());
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        TxnMetadata that = (TxnMetadata) o;
+
+        if (producerId != that.producerId) return false;
+        if (!Objects.equals(firstOffset, that.firstOffset)) return false;
+        return Objects.equals(lastOffset, that.lastOffset);
```

Comment:
I avoid non final fields usage in equals/hashcode implementations. I should have left a comment in the code on why these methods were added. 
I tried to keep similar semantics like case classes in Scala as earlier. Afaik, it does all the fields eq(reference checks) as the default implementation. TxnMetadata is used as a value of TreeMap in ProducerStateManager. Currently, I do not see any usages in map#containsValue in that. So, I prefer to avoid implementing it for now as you suggested.

Reply:
+1 to removing `equals` and `hashCode` and updating the tests to check the fields directly. And yes, I realized you were maintaining the previous behavior, but it seems like a good idea to fix these small things if it's easy & safe to do.

Reply:
I left more specific comments regarding these changes, so resolving this one.